File: Day05/Part1/main.pl
use strict;
use warnings;

sub set_corresponding_hash {
    my (%dic, @numbers) = @_;

    # print numbers
    print("Numbers: ", $numbers[0], " ", $numbers[1], " ", $numbers[2], "\n");
    
    # print keys
    print("Keys: ");
    foreach my $key (keys %dic) {
        print($key, " ", $dic{$key}, " ");
    }

    foreach my $i ($numbers[0] .. $numbers[0] + $numbers[2]) {
        %dic = (%dic, "" . $i => "" . $numbers[2]);
        print(%dic{"" . $i}, "\n");
    }
}

## read file
#print "File location> ";
#my $file = <>;
#
#chomp($file);
#
#open(FH, '<', $file) or die $!;
#
#my $file_content = "";
#
#while (<FH>) {
#    $file_content .= $_;
#}
#
#close(FH);
#
## split file contentby lines
#my @lines = split('\n', $file_content);
#
## print lines
#foreach my $line (@lines) {
#    print($line, "\n");
#}

my @seeds = ('-1' => 'seeds');

my %seed_to_soil = ('-1' => 'seed to soil');

my %soil_to_fertilizer = ('-1' => 'soil to fertilizer');

my %fertilizer_to_water = ('-1' => 'fertilizer to water');

my %water_to_light = ('-1' => 'water to light');

my %light_to_temp = ('-1' => 'light to temp');

my %temp_to_humidity = ('-1' => 'temp to humidity');

my %humidity_to_location = ('-1' => 'humidity to location');

my @test = (10, 50, 46);

set_corresponding_hash(%seed_to_soil, @test);

foreach my $key (keys %seed_to_soil) {
    print($key, " ", $seed_to_soil{$key}, "\n");
}

#for (my $i = 0; $i <= $#r; $i++) {
#    print($test[$i], " ");
#}

print("\n");

File: Day05/Part1/main.py
with open(input("Filename >"), "r") as file:
    lines = file.readlines()

def compute_seed(seeds, numbers):
    computed = []
    changed = []

    for i in range(len(seeds)):
        changed.append(False)
        computed.append(seeds[i])

    for nbr in numbers:
        for j in range(len(seeds)):
            if (seeds[j] < nbr[1] + nbr[2]) and (seeds[j] >= nbr[1]):
                computed[j] = (seeds[j] - nbr[1]) + nbr[0]
                changed[j] = True

    for i in range(len(seeds)):
        if not changed[i]:
            computed[i] = seeds[i]

    return computed

def get_nearest(numbers):
    nearest = numbers[0]

    for i in range(len(numbers)):
        if numbers[i] < nearest:
            nearest = numbers[i]

    return nearest

seeds = []
seeds_computed = []

dic = {}

seeds_nums = lines[0].replace("\n", "").split(" ")
seeds_nums.pop(0)

for i in range(len(seeds_nums)):
    seeds.append(int(seeds_nums[i]))

lines.pop(0)
lines.pop(0)
lines.pop(0)

print("Seeds\t\t\t\t", seeds)

i = 0

nums = []
while lines[i] != "\n":
    nbrs = lines[i].replace("\n", "").split(" ")
    nums.append([int(nbrs[0]), int(nbrs[1]), int(nbrs[2])])
    i += 1

seeds_computed = compute_seed(seeds, nums)
print("Seeds computed for soil\t\t", seeds_computed)

i = i + 2

nums = []
while lines[i] != "\n":
    nbrs = lines[i].replace("\n", "").split(" ")
    nums.append([int(nbrs[0]), int(nbrs[1]), int(nbrs[2])])
    i += 1

seeds_computed = compute_seed(seeds_computed, nums)
print("Seeds computed for fertilizer\t", seeds_computed)

i = i + 2

nums = []
while lines[i] != "\n":
    nbrs = lines[i].replace("\n", "").split(" ")
    nums.append([int(nbrs[0]), int(nbrs[1]), int(nbrs[2])])
    i += 1

seeds_computed = compute_seed(seeds_computed, nums)
print("Seeds computed for water\t", seeds_computed)

i = i + 2

nums = []
while lines[i] != "\n":
    nbrs = lines[i].replace("\n", "").split(" ")
    nums.append([int(nbrs[0]), int(nbrs[1]), int(nbrs[2])])
    i += 1

seeds_computed = compute_seed(seeds_computed, nums)
print("Seeds computed for light\t", seeds_computed)

i = i + 2

nums = []
while lines[i] != "\n":
    nbrs = lines[i].replace("\n", "").split(" ")
    nums.append([int(nbrs[0]), int(nbrs[1]), int(nbrs[2])])
    i += 1

seeds_computed = compute_seed(seeds_computed, nums)
print("Seeds computed for temperature\t", seeds_computed)

i = i + 2

nums = []
while lines[i] != "\n":
    nbrs = lines[i].replace("\n", "").split(" ")
    nums.append([int(nbrs[0]), int(nbrs[1]), int(nbrs[2])])
    i += 1

seeds_computed = compute_seed(seeds_computed, nums)
print("Seeds computed for humidity\t", seeds_computed)

i = i + 2

nums = []
while i < len(lines):
    nbrs = lines[i].replace("\n", "").split(" ")
    nums.append([int(nbrs[0]), int(nbrs[1]), int(nbrs[2])])
    i += 1

seeds_computed = compute_seed(seeds_computed, nums)
print("Seeds computed for location\t", seeds_computed)

print("Nearest:", get_nearest(seeds_computed))